### PR TITLE
[BugFix] Fixes #20983 - Store py_func dtypes for correct conversion

### LINKED ISSUE #20983: [Bug] tf.py_func : Bad tensor -> np.array conversion in py_func (return [] with dtype=int64 is returned as dtype=float64)
### Issue template:

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes (minimal example attached inline)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac OSX 10.12.6
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: Probably
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: ('v1.5.0-0-g37aa430d84', '1.5.0')
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: `python py_func_failure.py`

### Describe the problem

This bug occurs when a `tf.py_func` returns an empty python list `[]` which is intended to be a tensor of type `tf.int64` (as defined in the `Tout=` of the `py_func`). 

In the `ops.script_ops.py` the `_convert` static method is unable to tell the `numpy dtype` of the incoming tensor when it is an empty list.  This causes the `_convert` method to execute a `np.asarray([], dtype=None, order="C")` which gives us a `array([], dtype=float64)` instead of a `array([], dtype=int64)` which then does not agree with the output tensor description in the py_func. 

### Source code / logs

See gist here for the example that reproduces this issue: https://gist.githubusercontent.com/sabhiram/3f5eaf7e566ef9aefb3ae6e5b8d2edb0/raw/182ab6bc72b5b7b13fd55964c72030cbc53f7cb3/py_func_failure.py

Inlined here:
```
import tensorflow as tf

def test_func(x):
    """ Builds a list of ints with length `x`.
    """
    return [i for i in range(x)],


def main():
    t0 = tf.constant(0, dtype=tf.int64)
    t0 = tf.py_func(test_func, [t0], tf.int64)

    t1 = tf.constant(1, dtype=tf.int64)
    t1 = tf.py_func(test_func, [t1], tf.int64)

    with tf.Session() as sess:
        sess.run(t1)            # OK
        sess.run(t0)            # 0-th value is double, expects int64

if __name__ == "__main__":
    print(tf.GIT_VERSION, tf.VERSION)
    main()
```

### PR DESCRIPTION
**Pull request by sabhiram**
1.  Implement wrapper class `PythonFunc` to store Tout list
    for appropriate dtype conversion from tensorflow types to
    numpy types using the `dtype.as_numpy_dtype` member.
2.  Add py_func_test to illustrate broken behavior (see
    `testConvertEmptyList`).

### CODE REVIEW

**Comment by sabhiram** [comment]
^ WIP on the CI errors - they were due to the `FuncRegistry._convert()` being used from `dataset_ops`.  If this is being called this way, should convert be exposed from `script_ops` more officially (not through the FuncRegistry)?

My fixes are done, I am in-tow for the 9 million bazel files that needed recompilation :) Will squash and push once I confirm things are resolved locally.

File: tensorflow/python/ops/script_ops_test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import dtypes
+from tensorflow.python.ops import script_ops
+from tensorflow.python.platform import test
+
+class PyFuncTest(test.TestCase):
+  """Tests for script_ops.py_func."""
+
+  def test_basic_convert(self):
+    def get_not_empty_list(args):
```
Comment (akshayka) [inline tensorflow/python/ops/script_ops_test.py]: Can you delete this file and add your tests to tensorflow/python/kernel_tests/py_func_test.py?

File: tensorflow/python/BUILD [hunk 2]
```diff
@@ -2524,6 +2524,18 @@ py_library(
     ],
 )
 
+py_test(
```
Comment (akshayka) [inline tensorflow/python/BUILD]: Delete this build rule and instead add your tests to tensorflow/python/kernel_tests/py_func_test.py

Comment (sabhiram) [reply tensorflow/python/BUILD]: Can do - thanks for the pointers. Will fix and push shortly.

**Comment by sabhiram** [comment]
@akshayka - Thanks for the feedback, I think I should have done justice to your comments. Please let me know if this looks better. 

PS: One of the basic tests that I had originally added seemed useless given the py_func_test.py file, so I went ahead and removed it (preserved the test that will *fail* without these changes).

File: tensorflow/python/ops/script_ops.py [hunk 3]
```diff
@@ -171,8 +160,48 @@ def _convert(value, dtype=None):
...
+    if isinstance(ret, six.binary_type):
+      ret = [ret]
+    # Ensures that we return either a single numpy array or a list of numpy
+    # arrays.
+    if isinstance(ret, (tuple, list)):
+      return [
+          self._convert(x, dtype=dtype.as_numpy_dtype)
+          for (x, dtype) in zip(ret, self._out_dtypes)
+      ]
+    return self._convert(ret)  # TODO: pass self._out_dtypes?
```
Comment (akshayka) [inline tensorflow/python/ops/script_ops.py]: Is there any reason to not pass out_dtypes here?

Comment (sabhiram) [reply tensorflow/python/ops/script_ops.py]: One thing - the registry will force the Tout to be an `array / list` -- see ~line 270 in the `_internal_py_func` where the `Tout` is initialized.  So we would need to sanitize the _out_dtypes. But that should be easy.

**Comment by akshayka** [comment]
Great, go ahead and delete the test you commented out, I agree with you that it no longer applies.

**Comment by sabhiram** [comment]
@akshaym - any reservations?

**Comment by akshaym** [comment]
@sabhiram not from me. I think @akshayka is the best person to review this, so I'll unassign myself.

**Comment by sabhiram** [comment]
@akshayka - I think this is good to go - might need your blessing before the bot takes over.

**Review by akshayka — APPROVED** [review]

**Comment by sabhiram** [comment]
@akshayka - Heh, what did I break? Unfortunately I cannot peer into the CI process enough to guess how this affected what.  Happy to help see it through if needed.

**Comment by akshayka** [comment]
Hey Sabhiram, sorry for the delay. I didn't have the time to diagnose the issue, but the change caused an internal-only test to segfault.

**Pull request merged**